bjoern/request.c

```diff
@@ -121,50 +121,70 @@ on_path(http_parser* parser, const char* path, size_t len)
 {
   if(!(len = unquote_url_inplace((char*)path, len)))
     return 1;
-  _set_header_free_value(_PATH_INFO, _Unicode_FromStringAndSize(path, len));
+  _set_or_append_header(_PATH_INFO, path, len);
   return 0;
 }
 
 static int
 on_query_string(http_parser* parser, const char* query, size_t len)
 {
-  _set_header_free_value(_QUERY_STRING, _Unicode_FromStringAndSize(query, len));
+  _set_or_append_header(_QUERY_STRING, query, len);
   return 0;
 }
 
 static int
 on_header_field(http_parser* parser, const char* field, size_t len)
 {
-  if(PARSER->value.data) {
-    /* Store previous header and start a new one */
-    _set_header_from_http_header();
-  } else if(PARSER->field.data) {
-    UPDATE_LENGTH(field);
+  if(PARSER->last_was_data) {
+    _free_and_unset_if_set(PARSER->field);
+    PARSER->field = _Unicode_FromStringAndSize("HTTP_", 5);
+    PARSER->last_was_data = false;
+    PARSER->invalid_header = false;
+  }
+
+  if(PARSER->invalid_header) {
     return 0;
   }
-  PARSER->field = (string){(char*)field, len};
-  PARSER->value = (string){NULL, 0};
+
+  char field_processed[len];
+  for(size_t i=0; i<len; i++) {
+    char c = field[i];
+    if(c == '_') {
+      // CVE-2015-0219
+      PARSER->invalid_header = true;
```

Comment:
What do you think about simply aborting request parsing entirely in this case and responding with HTTP 400?

Reply:
This CVE is about django, where I found this: https://github.com/django/django/commit/316b8d49746933d1845d600314b002d9b64d3e3d#diff-f6d1c75ec606389da5af6558bf57f171R142

So they are only dropping this headers, but not the whole request.
Also the previous implementation was (https://github.com/jonashaag/bjoern/blob/master/bjoern/request.c#L250) that `wsgi_http_header` returns NULL, which was called by a macro on L103, and checked on the next line, and was ignored. So if I understand right, I didn't change the behaviour there, and as this is implemented in this way in other projects, i think it is a good idea to follow this. What do you think?

Reply:
Agreed.